lib/util/http/cache/abstract-http-cache-provider.ts

```diff
@@ -0,0 +1,97 @@
+import { logger } from '../../../logger';
+import { HttpCacheStats } from '../../stats';
+import type { GotOptions, HttpResponse } from '../types';
+import { copyResponse } from '../util';
+import { HttpCacheSchema } from './schema';
+import type { HttpCache, HttpCacheProvider } from './types';
+
+export abstract class AbstractHttpCacheProvider implements HttpCacheProvider {
+  protected abstract load(url: string): Promise<unknown>;
+  protected abstract persist(url: string, data: HttpCache): Promise<void>;
+
+  async get(url: string): Promise<HttpCache | null> {
+    const cache = await this.load(url);
+    const httpCache = HttpCacheSchema.parse(cache);
+    if (!httpCache) {
+      return null;
+    }
+
+    return httpCache as HttpCache;
+  }
+
+  async setCacheHeaders<T extends Pick<GotOptions, 'headers'>>(
+    url: string,
+    opts: T,
+  ): Promise<void> {
+    const httpCache = await this.get(url);
+    if (!httpCache) {
+      return;
+    }
+
+    opts.headers ??= {};
+
+    if (httpCache.etag) {
+      opts.headers['If-None-Match'] = httpCache.etag;
+    }
+
+    if (httpCache.lastModified) {
+      opts.headers['If-Modified-Since'] = httpCache.lastModified;
+    }
+  }
+
+  async wrapResponse<T>(
+    url: string,
+    resp: HttpResponse<T>,
+  ): Promise<HttpResponse<T>> {
+    if (resp.statusCode === 200) {
+      const etag = resp.headers?.['etag'];
+      const lastModified = resp.headers?.['last-modified'];
+
+      HttpCacheStats.incRemoteMisses(url);
+
+      const httpResponse = copyResponse(resp, true) as unknown as Record<
+        string,
+        unknown
+      >;
```

Comment:
Why this ugly cast?

Reply:
Never mind, just realized I've already made changes but forgot to remove type cast